File: packages/core/src/structured-fields/signature.ts
import type { ParsedSignature, ParsedSignatures } from '../types/public.js'
import { StructuredFieldParseError, splitDictionaryMembers } from './shared.js'

const LABEL_PATTERN = /^[A-Za-z*][A-Za-z0-9_.*-]*$/

/**
 * Parses a Signature header into a label-keyed dictionary of byte sequences.
 */
export const parseSignature = (headerValue: string): ParsedSignatures => {
  const members = splitDictionaryMembers(headerValue)

  if (members.length === 0) {
    throw new StructuredFieldParseError('The Signature header was empty.')
  }

  const parsed: ParsedSignatures = {}

  for (const member of members) {
    const equalsIndex = member.indexOf('=')

    if (equalsIndex === -1) {
      throw new StructuredFieldParseError(`Missing "=" in Signature member "${member}".`)
    }

    const label = member.slice(0, equalsIndex).trim()
    const rawValue = member.slice(equalsIndex + 1).trim()
    const match = rawValue.match(/^:([^:]+):$/)

    if (!LABEL_PATTERN.test(label)) {
      throw new StructuredFieldParseError(`Invalid Signature label "${label}".`)
    }

    if (!match) {
      throw new StructuredFieldParseError(
        `Signature member "${label}" must be encoded as an RFC 8941 byte sequence.`
      )
    }

    const value = match[1]

    if (!value) {
      throw new StructuredFieldParseError(`Signature member "${label}" contained an empty value.`)
    }

    parsed[label] = {
      label,
      raw: member,
      value
    } satisfies ParsedSignature
  }

  return parsed
}

File: languages/python/src/open_payments_http_signatures_devkit/signature.py
from __future__ import annotations

import re

from .errors import StructuredFieldParseError
from .signature_input import split_dictionary_members
from .types import ParsedSignature, ParsedSignatures

LABEL_PATTERN = re.compile(r"^[A-Za-z*][A-Za-z0-9_.*-]*$")


def parse_signature(header_value: str) -> ParsedSignatures:
    members = split_dictionary_members(header_value)

    if not members:
        raise StructuredFieldParseError("The Signature header was empty.")

    parsed: ParsedSignatures = {}

    for member in members:
        equals_index = member.find("=")

        if equals_index == -1:
            raise StructuredFieldParseError(f'Missing "=" in Signature member "{member}".')

        label = member[:equals_index].strip()
        raw_value = member[equals_index + 1 :].strip()
        match = re.fullmatch(r":([^:]+):", raw_value)

        if not LABEL_PATTERN.fullmatch(label):
            raise StructuredFieldParseError(f'Invalid Signature label "{label}".')

        if not match:
            raise StructuredFieldParseError(
                f'Signature member "{label}" must be encoded as an RFC 8941 byte sequence.'
            )

        value = match.group(1)

        if not value:
            raise StructuredFieldParseError(
                f'Signature member "{label}" contained an empty value.'
            )

        parsed[label] = ParsedSignature(label=label, value=value, raw=member)

    return parsed
